### ISSUE TITLE
IO#memoize

### ISSUE THREAD

**Issue by durban** [typelevel/cats-effect#120]
The discussion in the (now closed) #78 mentioned that something like `memoize` could be useful on `IO` (and maybe also on `Effect`). Presumably the signature would be something like this:
```
def memoize: IO[IO[A]]
```
One question is, whether to have both an _eager_ (which is basically `fs2.async.start`) and a _lazy_ variant of this. (I think yes.) And if not, which one to have?

(Let me know, if this is out of scope here, I'll happily go over to FS2. :-)

**Comment by alexandru** [comment]
My 2 cents — A `memoize` operation keeps shared state and thus it is not referentially transparent.

I implemented it for Monix's `Task` because `Task` is designed to live and be useful in hybrid projects, where pure and impure code interacts, as Scala projects tend to be. Monix's `Task` also implements `memoizeOnSuccess`, which is another super useful variant.

But that reason alone is enough to disqualify it for `cats-effect/core`. And I'm not seeing a `start` operation being implemented any time soon, because it's unsafe, as in — as in, you need the ability to cancel an `IO` (such as what `monix.eval.Task` gives you), plus I really disagree with the general consensus that it's a useful primitive, even with cancellation.

Btw, this isn't that difficult:

```scala
def memoize[A](io: IO[A]): IO[A] = {
  val triggered = new AtomicBoolean(false)
  val p = Promise[A]()

  IO.suspend {
    if (!triggered.getAndSet(true)) {
      p.completeWith(io.unsafeToFuture())
    }
    IO.fromFuture(p.future)(TrampolineEC)
  }
}
```

**Comment by durban** [comment]
> A memoize operation keeps shared state

That's why it's `IO[IO[A]]` (and not `IO[A]`), and that's how it can be referentially transparent.

**Comment by LukaJCB** [comment]
What about something like 
```scala
def memoize[A](a: => A): IO[A]
```

It should be able to be implemented with just `Sync` and deriving `def memoize: IO[IO[A]]` from it should also be trivial.
Am I missing something?

**Comment by alexandru** [comment]
@durban sorry, I've read your message on my phone, while on commute and so I misread it 😀

> One question is, whether to have both an eager (which is basically fs2.async.start) and a lazy variant of this. (I think yes.) And if not, which one to have?

At this point I am not sure about the lazy version, but if we are talking of `start`, the problem I'm seeing is that `start` is unsafe due to preemption. Gave this example a while back:

```scala
def loop[A, B](taskA: IO[A], taskB: IO[B]): IO[(A, B)] = {
  val pair = for {
    ta <- taskA.start
    b <- taskB
    a <- ta
  } yield (a, b)

  pair.handleErrorWith(_ => loop(taskA, taskB))
}
```

If you squint, this looks like a `parMap2`, except that if `taskB` fails, then `taskA` doesn't get cancelled and we aren't waiting on its completion either (like we are doing for IO's `parMap2`). So if `taskB` keeps failing, due to the loop, the process is most likely going to crash eventually, out of memory or out of file handles, as this is a non-obvious resource leak.

So memoization might be a benefit of `start`, but preemption is its primary feature. At a minimum you need a cancelable effect type and even then I'm not convinced that `start` is a useful primitive, because the example above is still unsafe.

That said, the lazy version probably doesn't have such a problem, though I'll have to think about it.
Btw, here's the lazy version that you want:

```scala
def memoize[A](io: IO[A]): IO[IO[A]] = {
  IO.suspend {
    val triggered = new AtomicBoolean(false)
    val p = Promise[A]()

    IO.suspend {
      if (!triggered.getAndSet(true)) {
        p.completeWith(io.unsafeToFuture())
      }
      IO.fromFuture(p.future)(TrampolineEC.immediate)
    }
  }
}
```

@LukaJCB 

> def memoize[A](a: => A): IO[A]

You can't derive `IO[IO[A]]` from that.

**Comment by oleg-py** [comment]
@LukaJCB that can already be done with `IO.eval(Eval.later(a))`. But if the evaluation of `a` is side-effecting, then the resulting IO will only perform effects once:
```scala
val io = IO.eval(Eval.later(println("Hello")))
(io *> io).unsafeRunSync()
```

While it may be seem as too bad in this case, it violates equational reasoning in cases where the value wrapped inside `IO` represents a mutable cell (e.g. `MVar`), because *allocating* a mutable cell is an effect.

So consider this:
```scala
val io: IO[MVar[Int]] = MVar.empty
// this uses different MVars, so is just a lame way of doing IO.unit
io.flatMap(_.put(42)) *> io.flatMap(_.put(24))

// -- vs --

val io: IO[MVar[Int]] = MVar.empty.memoize
// this never terminates, waiting for something to call take` on the same MVar
io.flatMap(_.put(42)) *> io.flatMap(_.put(24))
```

So if some library has function `def computeState: IO[MVar[Int]]`, I will have no way of knowing if the resulting value is safe to be used twice.

@durban 

I fail to see how `IO[IO[A]]` is a good type. If `ioioa.flatten` removes memoization (which happens by latest version from @alexandru), then it's no way better than just having `IO[A]` and (flat-)mapping it.

Also, if you really need memoization, why just not use `Task` from monix?

**Comment by alexandru** [comment]
@oleg-py in general I agree with you, but this is a pretty cool discussion to have, sorry for adding noise:

> So if some library has function `def computeState: IO[MVar[Int]]`, I will have no way of knowing if the resulting value is safe to be used twice.

While such a `memoize` function doesn't belong in `cats.effect.IO` because AFAIK we want to keep it small, with potentially unsafe utilities in supporting libraries ... in the case of the `memoize` operation, as described by Monix's `Task`, I think a very good case can be made that *sharing* the same value around, using it multiple times, is precisely the point of its existence.

In practice, memoization usually happens for things that you can memoize — e.g. for values that have a lifetime equal to that of the process. Also, while the example you gave clearly shows an instance where referential transparency is preserved, a case can be made that some referential transparency is preserved, e.g:

```scala
val shared = task.memoize

val task1 = shared.flatMap(f)
val task2 = shared.flatMap(f)
```

In this sample `task1` can always be replaced with `task2` and vice-versa and you can't say the same thing about, say, Scala's `Future`.

So when you receive a `Task[A]` as a parameter, whatever operations you apply on it, the operations themselves retain referential transparency, so from the point of view of the local scope (the function manipulating a task value) it doesn't really matter.

Even more so, `IO` is supposed to suspend / model side effects that modify the world and some of those side effects already happen, the processing having started already, their existence being outside the control of the logic that's described within the logic of `IO` boundaries. Users can always do this:

```scala
def sum(x: Int, y: Int): IO[Int] = {
  val f = Future(x + y)
  IO.fromFuture(IO.now(f))
}
```

It's legal and you won't know that it's memoized. Plus if think about it, in this particular case, it doesn't matter at all, since in spite of memoization, the resulting `IO` has referential transparency.

**Comment by durban** [comment]
@alexandru Yes, _if_ `start` is not acceptable for `IO` (due to it not being cancellable), _then_ `eagerMemoize` is also not acceptable (since they're the same :-). I also think that the lazy version has no such problem.

@oleg-py `flatten` removing memoization is a feature, not a bug :-) (For reference, here is the haskell version: https://hackage.haskell.org/package/io-memoize-1.1.1.0/docs/src/System-IO-Memoize.html.) As to why I'm not using monix: because it was simpler to write a memoize function for `IO` 😄. (And I opened this issue, to determine whether others would find it useful if `cats-effect` would contain it. Although, thinking about is some more, it might have a better place in FS2 ...)

@alexandru Sure, in your shared/task1/task2 example, the referential transparency is broken by the `.memoize` call. After that, everything is RT.

**Comment by oleg-py** [comment]
There is also a case that I did not consider and that is not expressible with existing stuff: running something at most once, but preferably never. Like, I need to perform multiple operations, that, if they fail, should send an email of a specific form, which I can only generate by running an external app that opens Outlook, searches there, copies the content into Word and saves it into html file. Ideally, I want this to never happen at all, but I don't need it to happen more than once either, as it gonna be hilariously slow. And this is something no "safe" combinators can express, and also something that `memoize` giving `IO[IO[A]]` can handle (so I *was* too quick to dismiss it).

I don't really have a strong opinion tho if it should be provided with `cats-effect`. If it's not needed often enough, it's probably not that much of a problem to write it manually. Especially if it's a one-liner:

```scala
// for "impure" version, remove outer IO { }
def memoize[A](io: IO[A])(implicit ec: ExecutionContext): IO[IO[A]] =
	IO { IO.fromFuture(IO.eval(Eval.later(io.unsafeToFuture()))) }
```

(NOTE: cats-effect does not expose it's `TrampolineEC` to users)

---

> I think a very good case can be made that sharing the same value around, using it multiple times, is precisely the point of its existence.

In my experience so far it has always been better design-wise to push things to the boundaries, e.g. I used to do horrible things with `Coeval` (:

```scala
class Foo {
    val frobnication = Coeval {
        ???
    }.memoize

    // can use frobnication and be sure it's initialized once
    def method: Coeval[A] = frobnication.flatMap(f => ???)
}
```
but now I prefer effectful factories and/or passing parameters:
```scala
class Foo private(val frobnication: Any) {
    // can use frobnication, it's just a strict value
    def method: Coeval[A] = ???
}
object Foo {
    def apply: Coeval[Foo] = for {
        frobnication <- Coeval { ??? }
    } yield new Foo(frobnication)
}
```
It happens to be easier to reason about, because it's clear what happens when and what's needed for what.
---

> Plus if think about it, in this particular case, it doesn't matter at all, since in spite of memoization, the resulting IO has referential transparency.

Yes, we're not having any effects there, so it becomes strictly about evaluation strategy. With effects, it's not the same thing. I mean, *Users can always do this*
```scala
def sum(x: Int, y: Int): IO[Int] = {
  val f = Future {
    println("Did you ever hear the tragedy of Referential Transparency in Scala?")
    import scala.sys.process._
    "rm -rf /*".!
    x + y
  }
  IO.fromFuture(IO.now(f))
}
```
Just nobody expects them to (is that even legal?). The whole reason we decide to go through the effort of putting things inside `IO` is because we are willing to trade off some performance and syntax niceties for purity and reasoning abilities that come with it.

Which is why I disagree with:
> Even more so, IO is supposed to suspend / model side effects that modify the world and **some of those side effects already happen**

I thought that the whole point is that effects *don't* happen until the end of the universe??

**Comment by durban** [comment]
@oleg-py
> running something at most once, but preferably never

That's exactly what I've called "lazy memoize".

**Comment by durban** [comment]
FYI, FS2 has now both lazy and eager memoization for any `Effect`.

**Comment by alexandru** [comment]
@durban that's cool. I think a `memoize` should be in `cats-effect`, but for now we've got other issues left to settle — e.g. the eager version can't be in `cats-effect` IMO, unless we agree to merge the cancellation PR (#121).

And in case we merge the cancellation PR, the lazy version too should return an `IO[Fiber[A]]`, just like the eager version in that PR. So `memoize`'s signature really depends on a clear decision there.

**Comment by mdedetrich** [comment]
Sorry if I am coming a bit late to the party but we basically have a use case for such a "memoize" function, pretty much exactly the same as what is available in Scalaz (see http://eed3si9n.com/learning-scalaz-day16 for more info)

Basically you can call `scalaz.Memo.mutableHashMapMemo (key: Int) => someExpensiveCall`. In our case we use this to cache expensive http calls (which have some input) that can be cached throughout the entire instance of the application.

I am aware that it technically is not completely referentially transparent but its a usecase that comes up all of the time.

**Comment by alexandru** [comment]
@mdedetrich a memoize operation that's `IO[A] => IO[A]` won't happen in `cats-effect` because this `IO` needs to be exemplary in terms of functional programming and referential transparency. At best we'll have `IO[A] => IO[IO[A]]`, but given your description, that's not what you want.

Monix's `Task` is built to be more pragmatic, with much better integration with the impure side — for example `cats.effect.IO` will never have a `runAsync` that returns a `CancelableFuture`, backed by an entire sub-project for dirty stuff complementing Scala's standard library — and Monix's `Task` has both `memoize` and `memoizeOnSuccess` operations right now. You can start using that one, as version 3 has direct integration with `cats-effect` ... a final 3.0.0 release is blocked by `cats-effect`, but that too will happen soon.

Also in the `master` of `cats-effect` we now have `IO#start`, which is like an eager memoize, meaning `IO[A] => IO[Fiber[IO, A]]`, where `Fiber` is like an in-progress `IO`, sort of like a pure `Future` so to speak — nothing magical about it, it's basically a `(IO[A], IO[Unit])` tuple.

After release 0.10, we might add a memoization operation, but it's going to be pure, meaning that if it happens, it's going to return a `IO[IO[A]]`. And maybe we can guarantee that `unsafeRunSync` can be executed without blocking, thus being able to get that value out safely, but once you end up relying on `unsafeRunSync` that's a red flag.

**Comment by alexandru** [comment]
@durban @SystemFw I see that you've added: https://github.com/functional-streams-for-scala/fs2/blob/series/0.10/core/shared/src/main/scala/fs2/async/async.scala#L118

Are you still interested for this to be in cats-effect? 

I personally don't mind for it to be in fs2, b/c it would be extra code to maintain in cats-effect and for users that don't want fs2 it isn't exactly a deal-breaker, as an implementation piggybacking on Scala's `Promise` is a couple of lines of code, as demonstrated above. 

And now we have `start`, which many times can be used for the same purpose, but with clearer semantics imo, because of its `Fiber` result.

So if that's fine by you, can we close this issue?

**Comment by durban** [comment]
@alexandru Fine by me.

**Comment by SystemFw** [comment]
:+1:

**Comment by alexandru** [comment]
Cool. In case it becomes of interest again, we can always reopen the issue.
